cscs-checks/microbenchmarks/mpi/src/halo_cell_exchange.c

```diff
@@ -0,0 +1,156 @@
+                }
+                MPI_Bcast(&halosize[i], 1, MPI_INT, 0, MPI_COMM_WORLD);
+            }
+            if (j>comm_size){
+                if (comm_rank==0){
+                    printf("Please run with at least %d MPI ranks.\n", j);
+                }
+                ierr = MPI_Finalize();
+                if (ierr!=0) exit(1);
+                exit(0);
+            }
+
+            color = (comm_rank < j);
+            if (color == 0){
+                ierr = MPI_Comm_split(MPI_COMM_WORLD, MPI_UNDEFINED, comm_rank, &red_comm);
+            }else{
+                ierr = MPI_Comm_split(MPI_COMM_WORLD, color, comm_rank, &red_comm);
+                ierr = MPI_Cart_create(red_comm, ndims, dim_size, periods, reorder, &cart_comm);
+                if (ierr!=0) exit(1);
+                ierr = MPI_Comm_size(cart_comm, &comm_size_cart);
+                if (ierr!=0) exit(1);
+                ierr = MPI_Comm_rank(cart_comm, &comm_rank_cart);
+                if (ierr!=0) exit(1);
+
+                j=0;
+                for (i=0; i<ndims; i++){
+                    if (halosize[i]>j){
+                        j=halosize[i];
+                    }
+                }
+                sendbuf = (char*) malloc(ndims*2*j*sizeof(char));
+                recvbuf = (char*) malloc(ndims*2*j*sizeof(char));
+                request = (MPI_Request*) malloc(ndims*2*2*sizeof(MPI_Request));
+                status = (MPI_Status*) malloc(ndims*2*2*sizeof(MPI_Status));
+
+                start = MPI_Wtime ();
+
+                for (j=0; j<10000; j++){
+                    for (i=0; i<ndims; i++){
+                        ierr = MPI_Cart_shift(cart_comm, i, 1, &rank_source, &rank_dest);
+                        if (ierr!=0) exit(1);
+                        ierr = MPI_Irecv(recvbuf+i*2*halosize[i]*sizeof(char), halosize[i], MPI_CHAR, rank_source, 1, cart_comm, request+i*2);
+                        if (ierr!=0) exit(1);
+                        ierr = MPI_Irecv(recvbuf+(i*2+1)*halosize[i]*sizeof(char), halosize[i], MPI_CHAR, rank_dest, 1, cart_comm, request+i*2+1);
+                        if (ierr!=0) exit(1);
+                    }
+                    for (i=0; i<ndims; i++){
+                        ierr = MPI_Cart_shift(cart_comm, i, 1, &rank_source, &rank_dest);
+                        if (ierr!=0) exit(1);
+                        ierr = MPI_Isend(sendbuf+i*2*halosize[i]*sizeof(char), halosize[i], MPI_CHAR, rank_source, 1, cart_comm, request+i*2+ndims*2);
+                        if (ierr!=0) exit(1);
+                        ierr = MPI_Isend(sendbuf+(i*2+1)*halosize[i]*sizeof(char), halosize[i], MPI_CHAR, rank_dest, 1, cart_comm, request+i*2+1+ndims*2);
+                        if (ierr!=0) exit(1);
+                    }
+                    ierr = MPI_Waitall(ndims*2*2, request, status);
+                    if (ierr!=0) exit(1);
+                }
+
+                stop = MPI_Wtime ();
+                ttt = stop - start;
```

Comment:
Don't be so cryptical in variable names:

```suggestion
                elapsed_time = stop - start;
```